provider/src/main/kotlin/org/gradle/kotlin/dsl/codegen/GradleApiExtensions.kt

```diff
@@ -0,0 +1,438 @@
+ *
+ * GradleApiExtensions Limitations:
+ * - does not support generating extensions for functions with type parameters having multiple bounds (Kotlin where clause)
+ */
+internal
+fun writeGradleApiExtensionsTo(file: File, jars: Iterable<File>) {
+    jars.filter(gradleJarsFilter).toList().let { gradleJars ->
+        file.bufferedWriter().use {
+            it.apply {
+                write(fileHeader)
+                apiTypeProviderFor(gradleJars, parameterNamesSupplierFor(gradleJars)).use { api ->
+                    gradleApiExtensionDeclarationsFor(api).forEach {
+                        write("\n$it\n")
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+private
+val gradleJarsFilter = { jar: File ->
+    jar.name.startsWith("gradle-") && !jar.name.startsWith("gradle-kotlin-")
+}
+
+
+internal
+fun parameterNamesSupplierFor(gradleJars: List<File>) =
+    JarFile(gradleJars.single { it.name.startsWith("gradle-api-parameter-names-") }).use { jar ->
+        jar.getInputStream(jar.getJarEntry("gradle-api-parameter-names.properties")).use { input ->
+            Properties().also { it.load(input) }.let { index ->
+                { key: String ->
+                    index.getProperty(key, null)?.split(",")
+                }
+            }
+        }
+    }
+
+
+internal
+fun gradleApiExtensionDeclarationsFor(api: ApiTypeProvider): Sequence<String> =
+    api.allTypes().filter { it.isGradleApi }.flatMap { type ->
+        sequenceOf(reifiedTypeParametersExtensionsGenerator, kClassExtensionsGenerator)
+            .flatMap { generator -> generator(type) }
+            .distinctBy { it.signatureKey }
+            .map { it.toKotlinString() }
+    }
+
+
+private
+data class KotlinExtensionFunction(
+    val description: String,
+    val isIncubating: Boolean,
+    val isDeprecated: Boolean,
+    val isInline: Boolean,
+    val typeParameters: List<Pair<ApiTypeUsage, Boolean>>,
+    val targetType: ApiType,
+    val name: String,
+    val parameters: List<ApiFunctionParameter>,
+    val returnType: ApiTypeUsage,
+    val expressionBody: String
+) {
+
+    val signatureKey: Int = Objects.hash(targetType.sourceName, name, parameters.map { it.type })
+
+    fun toKotlinString(): String = StringBuilder().apply {
+        appendln("""
+            /**
+             * $description.
+             */
+        """.trimIndent())
+        if (isDeprecated) appendln("""@Deprecated("Deprecated Gradle API")""")
+        if (isIncubating) appendln("@org.gradle.api.Incubating")
+        if (isInline) append("inline ")
+        append("fun ")
+        if (typeParameters.isNotEmpty()) append("${typeParameters.joinInAngleBrackets { (if (it.second) "reified " else "") + it.first.toTypeParameterString() }} ")
+        append(targetType.sourceName)
+        if (targetType.typeParameters.isNotEmpty()) append(targetType.typeParameters.toTypeArgumentsString(targetType))
+        append(".")
+        append(name)
```

Comment:
Should the name have backticks surrounding it incase it's a keyword like `when`?